lib/FrontendTool/FrontendTool.cpp

```diff
@@ -219,38 +239,79 @@ static bool emitLoadedModuleTraceIfNeeded(ASTContext &ctxt,
     return true;
   }
 
-  llvm::SmallVector<std::string, 16> swiftModules;
+  ModuleDecl* currentModule = ctxt.getModuleByName(moduleName);
+  ModuleDecl::ImportFilter filter = ModuleDecl::ImportFilterKind::Public;
+  filter |= ModuleDecl::ImportFilterKind::Private;
+  filter |= ModuleDecl::ImportFilterKind::ImplementationOnly;
+  llvm::SmallVector<ModuleDecl::ImportedModule, 8> imports;
+  currentModule->getImportedModules(imports, filter);
+
+  llvm::StringSet<> importedModulePaths;
+  for (std::pair<ModuleDecl::AccessPathTy, ModuleDecl *> &import : imports)
+    importedModulePaths.insert(import.second->getModuleFilename());
+
+  llvm::DenseMap<StringRef, ModuleDecl *> pathToModuleDecl;
+  for (auto &module : ctxt.LoadedModules) {
+    ModuleDecl *loadedDecl = module.second;
+    assert(loadedDecl && "Expected loaded module to be non-null.");
+    pathToModuleDecl.insert(
+      std::pair<StringRef, ModuleDecl *>(loadedDecl->getModuleFilename(),
```

Comment:
You should handle the case where a module doesn't claim any filename. (The Builtin module, for example, although I'm not sure that ends up in LoadedModules. Also true for various LLDB modules.) Or if you don't want to handle it, you can assert it so we find out about it later.

Also, instead of calling `std::pair`'s constructor directly, you can use `std::make_pair`, or even just list initialization with `{}`.

Reply:
1. So the type of `getModuleFilename()` is not right, it should be `optional<string>`, returning empty strings is not very meaningful :(. Someone might make the same mistake I made right here. Should I submit a separate PR to fix it?

2. Thanks, I'll use `make_pair`.

Reply:
I'm not sure I agree about that. *Because* an empty string isn't a meaningful filename, it's a reasonable placeholder. Having it be `optional<string>` could lead to questions about what `some("")` would mean.

Reply:
Is there a standard `nonempty_string` newtype? We could use that. This [function](https://doc.rust-lang.org/src/std/path.rs.html#1974-1985) in the Rust stdlib for `parent` returns `Option<&Path>` instead of `&Path` with an empty path. For pointers, I'll probably get a segfault if I forget the nullptr case, but with strings, it is harder to catch a problem because you can do a lot more things with an empty string vs a nullptr.

Reply:
You could push for one, but we don't have one yet. I'll admit we're more blasé about using suboptimal types in C++ than we are in Swift, even though we could put more info into the C++ type system.